### PR TITLE
Run amq benchmark pod and get test results

### PR DESCRIPTION
**Pull request by Akarsha-rai**
Creates amq benchmark pods and get the test results. And also supports creation of multiple topics, partitions, producer, consumer pods.

Signed-off-by: Akarsha <akrai@redhat.com>

### CODE REVIEW

**Comment by Akarsha-rai** [comment]
Test passed: 
[Test_run.txt](https://github.com/red-hat-storage/ocs-ci/files/4747474/Test_run.txt)

File: ocs_ci/ocs/amq.py [hunk 1]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
 
-    # ToDo: Install helm and get kafka metrics
+    def run_amq_benchmark(
+        self, name="benchmark", kafka_namespace=constants.AMQ_NAMESPACE,
+        tiller_namespace="tiller", num_of_clients=8, worker=None, timeout=1200,
+        workload_name='1-topic-16-partition-1Kb', topics=1,
+        partitions_per_topic=16, message_size=1024, payload_file="payload/payload-1Kb.data",
+        subscriptions_per_topic=1, consumer_per_subscription=3, producers_per_topic=3,
+        producer_rate=50000, consumer_backlog_sizegb=0, test_duration_minutes=15
+    ):
```
Comment (hnallurv) [inline ocs_ci/ocs/amq.py:457]: there are too many parameters in this function call. can we put all these parameter in a dict and pass dict to function?

File: ocs_ci/ocs/amq.py [hunk 2]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+            self.crb_tiller = OCS(**sa_tiller[1])
+            self.sa_tiller.create()
+            self.crb_tiller.create()
+        except(CommandFailed, CalledProcessError) as cf:
+            log.error('Failed during creation of service account tiller')
+            raise cf
+
+        # Install helm cli (version v2.16.0 as we need tiller component)
+        # And create tiller pods
+        url = "https://get.helm.sh/helm-v2.16.1-linux-amd64.tar.gz"
```
Comment (hnallurv) [inline ocs_ci/ocs/amq.py:509]: what if the version of helm changes in future and we have to use most recent version? better not to hardcode it. check

File: ocs_ci/ocs/amq.py [hunk 3]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+                cwd=self.dir,
+                check=True
+            )
+        except (CommandFailed, CalledProcessError)as cf:
+            log.error('Error during installing helm')
+            raise cf
+
+        # Validate tiller pod is running
+        log.info("Waiting for 30s")
+        time.sleep(30)
```
Comment (hnallurv) [inline ocs_ci/ocs/amq.py:542]: Is 30 sec enough? should there be some re-tries?

File: ocs_ci/ocs/amq.py [hunk 4]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+
+        # Update commonConfig with kafka-bootstrap server details
+        driver_kafka = templating.load_yaml(constants.AMQ_DRIVER_KAFKA_YAML)
+        driver_kafka['commonConfig'] = (
+            f'bootstrap.servers=my-cluster-kafka-bootstrap.{kafka_namespace}.svc.cluster.local:9092'
+        )
+        json_file = f'{self.dir}/driver_kafka'
+        templating.dump_data_to_json(driver_kafka, json_file)
+        cmd = f'oc cp {json_file} {name}-driver:/'
+        run_cmd(cmd)
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:583]: Please use exec_cmd

File: ocs_ci/ocs/amq.py [hunk 5]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+        amq_workload_yaml['subscriptionsPerTopic'] = subscriptions_per_topic
+        amq_workload_yaml['consumerPerSubscription'] = consumer_per_subscription
+        amq_workload_yaml['producersPerTopic'] = producers_per_topic
+        amq_workload_yaml['producerRate'] = producer_rate
+        amq_workload_yaml['consumerBacklogSizeGB'] = consumer_backlog_sizegb
+        amq_workload_yaml['testDurationMinutes'] = test_duration_minutes
+        yaml_file = f'{self.dir}/amq_workload.yaml'
+        templating.dump_data_to_temp_yaml(amq_workload_yaml, yaml_file)
+        cmd = f'oc cp {yaml_file} {name}-driver:/'
+        run_cmd(cmd)
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:601]: here as well.

File: ocs_ci/ocs/amq.py [hunk 6]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+        # Create benchmark pods
+        log.info("Create benchmark pods")
+        try:
+            values = templating.load_yaml(constants.AMQ_BENCHMARK_VALUE_YAML)
+            values["numWorkers"] = num_of_clients
+            benchmark_cmd = (
+                f"linux-amd64/helm install {constants.AMQ_BENCHMARK_POD_YAML}"
+                f" --name {name} --tiller-namespace {tiller_namespace}"
+            )
+            run(
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:557]: Any reason why you are not using exec_cmds which are already present in ocs-ci?
This comment is for all the places where you are using run

File: ocs_ci/ocs/amq.py [hunk 7]
```diff
@@ -489,27 +653,47 @@ def setup_amq_cluster(self, sc_name, namespace=constants.AMQ_NAMESPACE, size=100
         self.amq_is_setup = True
         return self
 
-    def cleanup(self, namespace=constants.AMQ_NAMESPACE):
+    def cleanup(self, kafka_namespace=constants.AMQ_NAMESPACE, tiller_namespace='tiller'):
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:656]: Please create a constant for triller in constants.py and use the constant here

File: ocs_ci/ocs/amq.py [hunk 8]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
 
         return threads
 
-    # ToDo: Install helm and get kafka metrics
+    def run_amq_benchmark(
+        self, name="benchmark", kafka_namespace=constants.AMQ_NAMESPACE,
+        tiller_namespace="tiller", num_of_clients=8, worker=None, timeout=1200,
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:452]: Please create a constant for triller in constants.py and use the constant here

File: ocs_ci/ocs/amq.py [hunk 9]
```diff
@@ -446,7 +447,170 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
 
         return threads
 
-    # ToDo: Install helm and get kafka metrics
+    def run_amq_benchmark(
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:450]: Please apply these changes to an existing test case and verify the PR

File: ocs_ci/ocs/amq.py (hunk 2 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:509]: From the doc https://gist.github.com/ksingh7/31da820f522050ab9e0a7c1599d13d6b, it is the stable one so used the same version

[the remark this answers is not in the document]
Comment (shreekarSS) [reply ocs_ci/ocs/amq.py:509]: url = "https://get.helm.sh/helm-v2.16.1-linux-amd64.tar.gz" --> can be moved to constats

**Comment by Akarsha-rai** [comment]
Observed sometimes even benchmark pod didn't run, test-case would have passed so added validation and parsing of the result in same PR.

Locally ran and verified. 
[testrun.txt](https://github.com/red-hat-storage/ocs-ci/files/4799670/testrun.txt)

File: ocs_ci/ocs/amq.py (hunk 9 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:452]: done

File: ocs_ci/ocs/amq.py (hunk 3 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:522]: In 30s pod would be up and in line523 it will wait till the pod to running state

File: ocs_ci/ocs/amq.py (hunk 1 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:457]: agreed, passed amq_workload_yaml as dict where all the amq workloads information should be updated

File: ocs_ci/ocs/amq.py (hunk 2 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: As discussed, will take it in next PR

File: ocs_ci/ocs/amq.py (hunk 4 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: done, so marking as resolved

File: ocs_ci/ocs/amq.py (hunk 5 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: done, so marking as resolved

File: ocs_ci/ocs/amq.py (hunk 7 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: As discussed, moved at the top of file

File: ocs_ci/ocs/amq.py (hunk 6 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: Changed to exec_cmds, so marking as resolved

File: ocs_ci/ocs/amq.py (hunk 8 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: As discussed, moved at the top of file

File: ocs_ci/ocs/amq.py (hunk 3 above)
Comment (PrasadDesala) [reply ocs_ci/ocs/amq.py:522]: Is waiting for 30sec needed? 
In line523 any how we are waiting for the tiller pod to reach running state.

File: ocs_ci/ocs/amq.py [hunk 10]
```diff
@@ -446,7 +449,194 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+        else:
+            cmd = "bin/benchmark --drivers /driver_kafka /amq_workload.yaml"
+        log.info(f"Run benchmark and running command {cmd} inside the benchmark pod ")
+        pod_obj = get_pod_obj(name=f"{benchmark_pod_name}-driver", namespace=tiller_namespace)
+        result = pod_obj.exec_cmd_on_pod(command=cmd, out_yaml_format=False, timeout=timeout)
+
+        self.benchmark = True
+
+        # Parse the json output
+        return self.validate_amq_benchmark(benchmark_pod_name, result, amq_workload_yaml)
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:576]: Any reason why we are returning the amq benchmark validation in the same run_benchmark function?
What if we want to run some step in the test, while amq_benchmark is in-progress?
Please let me know if I am missing anything here.

File: ocs_ci/ocs/amq.py [hunk 11]
```diff
@@ -489,27 +679,48 @@ def setup_amq_cluster(self, sc_name, namespace=constants.AMQ_NAMESPACE, size=100
         self.amq_is_setup = True
         return self
 
-    def cleanup(self, namespace=constants.AMQ_NAMESPACE):
+    def cleanup(self, kafka_namespace=constants.AMQ_NAMESPACE,
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:682]: break the line after (

File: ocs_ci/ocs/amq.py [hunk 12]
```diff
@@ -489,27 +679,48 @@ def setup_amq_cluster(self, sc_name, namespace=constants.AMQ_NAMESPACE, size=100
...
+                tiller_namespace=AMQ_BENCHMARK_NAMESPACE):
         """
         Clean up function,
         will start to delete from amq cluster operator
         then amq-connector, persistent, bridge, at the end it will delete the created namespace
 
         Args:
-            namespace (str): Created namespace for amq
+            kafka_namespace (str): Created namespace for amq
+            tiller_namespace (str): Created namespace for benchmark
```
Comment (PrasadDesala) [inline ocs_ci/ocs/amq.py:691]: Missing one empty line after this

File: tests/e2e/workloads/amq/test_amq_streamer_creation.py [hunk 13]
```diff
@@ -46,15 +46,10 @@ def test_install_amq_cephfs(self, interface, test_fixture_amq):
...
-        # Check messages are sent and received
-        threads = test_fixture_amq.run_in_bg()
-        for t in threads:
-            t.join()
+        # Run benchmark
+        amq_workload_dict = templating.load_yaml(constants.AMQ_WORKLOAD_YAML)
+        amq_workload_dict['producersPerTopic'] = 3
+        amq_workload_dict['consumerPerSubscription'] = 3
+        assert test_fixture_amq.run_amq_benchmark(amq_workload_yaml=amq_workload_dict) is not None, (
+            "Failed to run amq benchmark pod"
```
Comment (PrasadDesala) [inline tests/e2e/workloads/amq/test_amq_streamer_creation.py:54]: You are also doing the validation in the same run_amq_benchmark method

File: ocs_ci/ocs/amq.py (hunk 10 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:576]: Sometimes workload wouldn't have run completely or might have failed in between but the test case showed as passed. So added the validation on same function. 

In coming PR will pass `validate_amq_benchmark` as param(bool type).

File: ocs_ci/ocs/amq.py [hunk 14]
```diff
@@ -446,7 +449,194 @@ def run_in_bg(self, namespace=constants.AMQ_NAMESPACE, value='10000', since_time
...
+            cmd = f"bin/benchmark --drivers /driver_kafka --workers {worker} /amq_workload.yaml"
+        else:
+            cmd = "bin/benchmark --drivers /driver_kafka /amq_workload.yaml"
+        log.info(f"Run benchmark and running command {cmd} inside the benchmark pod ")
+        pod_obj = get_pod_obj(name=f"{benchmark_pod_name}-driver", namespace=tiller_namespace)
+        result = pod_obj.exec_cmd_on_pod(command=cmd, out_yaml_format=False, timeout=timeout)
+
+        self.benchmark = True
+
+        # Parse the json output
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:575]: Thank you, will change according:)

File: ocs_ci/ocs/amq.py (hunk 3 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:522]: No, it takes time to pod to spin and in next line we are waiting till the pod to reach running

File: ocs_ci/ocs/amq.py (hunk 12 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py:694]: done

File: ocs_ci/ocs/amq.py (hunk 10 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/amq.py]: As discussed, removed validating in this function

File: tests/e2e/workloads/amq/test_amq_streamer_creation.py [hunk 15]
```diff
@@ -46,15 +46,11 @@ def test_install_amq_cephfs(self, interface, test_fixture_amq):
...
-        # Wait for some time to generate msg
-        waiting_time = 60
-        log.info(f"Waiting for {waiting_time}sec to generate msg")
-        time.sleep(waiting_time)
-
-        # Check messages are sent and received
-        threads = test_fixture_amq.run_in_bg()
-        for t in threads:
-            t.join()
+        # Run benchmark
```
Comment (hnallurv) [inline tests/e2e/workloads/amq/test_amq_streamer_creation.py:49]: As per https://docs.google.com/presentation/d/15WV0AmtTy58G2JRM8fAfYM6hVDl3oljrg8eKEswKHBk/edit#slide=id.g70daec23bd_0_312, AMQ is recommended on CephFS. Better to ignore running on cephfs for now.

Comment (Akarsha-rai) [reply tests/e2e/workloads/amq/test_amq_streamer_creation.py:46]: Updated a PR to run amq on rbd only

**Pull request merged**